Sources/ServiceLifecycle/Service.swift

```diff
@@ -0,0 +1,51 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftServiceLifecycle open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftServiceLifecycle project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftServiceLifecycle project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// This is the basic protocol that a service has to implement.
+public protocol Service: Sendable {
+    /// This indicates if the service is expected to be running for the entire duration.
+    /// If a long running service is returning from the ``Service/run()-1ougx`` method it is treated
+    /// as an unexpected early exit and all other services are going be cancelled.
+    ///
+    /// By default this is set to `true`.
+    var isLongRunning: Bool { get }
+
+    /// This method is called when the ``ServiceRunner`` is starting all the services.
+    ///
+    /// Concrete implementation should execute their long running work in this method such as:
+    /// - Handling incoming connections and requests
+    /// - Background refreshes
+    func run() async throws
+
+    /// This method is called when the ``ServiceRunner`` receives a graceful shutdown signal.
+    ///
+    /// Concrete implementations can implement this if they support graceful shutdown.
+    func shutdownGracefully() async throws
```

Comment:
Task cancellation and `shutdownGracefully` serve two entirely different purposes. Task cancellation is a _forceful_ action where every child task end their work as soon as possible, e.g. most `AsyncSequence`s are returning `nil` or throw `CancellationError` when the consuming task is cancelled.
The purpose behind `shutdownGracefully` is to allow (mostly) servers to implement quiescing logic. While we could just cancel the task when we receive a graceful shutdown signal this would leave the users almost no way to implement any custom quiescing logic since their task just got cancelled.

Imagine the following scenario. We have a gRPC server that does bidirectional streaming of `Foo` objects. When we want to gracefully shutdown the the server our business logic wants to send out the latest state of `Foo` and a message indicating that the server is bouncing. If we would use task cancellation here then the task that is currently handling the bidirectional stream will just get cancelled. That's why we need a separate mechanism to inform the users of such an event. 

Tl;dr: task cancellation != graceful shutdown

Side note: It would be create if we could get something like structured event handling into the stdlib that allows for the same pattern as `withTaskCancellation` but for arbitrary events.

Reply:
Though we can discuss if we want to change the pattern to be `withShutdownGracefulHandler` instead of an explicit method. We can make that work outside the stdlib probably.

Reply:
Okay. Fine with me. Can we add this discussion GracefulShutdown vs. TaskCancellation as docu? And should users still implement Cancellation?